Framework/Kernel/src/Timer.cpp

```diff
@@ -60,4 +60,70 @@ std::ostream &operator<<(std::ostream &out, const Timer &obj) {
   return out;
 }
 
+//------------------------------------------------------------------------
+/** Instantiate the object and start the timer
+ * @param name :: custom name of the code block
+ * @param out :: stream to output the elapsed wall-clock time
+ */
+CodeBlockTimer::CodeBlockTimer(const std::string &name, std::ostream &out)
+    : name(name), out(out), start(std::chrono::system_clock::now()) {}
+
+/** Calculate and output the elapsed wall-clock time in seconds
+ */
+CodeBlockTimer::~CodeBlockTimer() {
+  const auto stop = std::chrono::system_clock::now();
+  const std::chrono::duration<double> elapsed = stop - start;
+  out << "Elapsed time (s) in \"" << name << "\": " << elapsed.count() << '\n';
+}
+
+//------------------------------------------------------------------------
+/** Instantiate the object and start the timer
+ * @param accumulator :: a persistent object keeping track of the total elapsed wall-clock time
+ */
+CodeBlockMultipleTimer::CodeBlockMultipleTimer(CodeBlockMultipleTimer::TimeAccumulator &accumulator)
+    : accumulator(accumulator), start(std::chrono::system_clock::now()) {}
+
+/** Calculate the elapsed wall-clock time (seconds) and update the time accumulator
+ */
+CodeBlockMultipleTimer::~CodeBlockMultipleTimer() {
+  const auto stop = std::chrono::system_clock::now();
+  const std::chrono::duration<double> elapsed = stop - start;
+  accumulator.increment(elapsed.count());
+}
+
+//------------------------------------------------------------------------
+/** Instantiate the object
+ * @param name :: custom name of the code block
+ */
+CodeBlockMultipleTimer::TimeAccumulator::TimeAccumulator(const std::string &name) : name(name) {}
+
+/** Reset the elapsed wall-clock time and number of times the code block was entered
+ */
+void CodeBlockMultipleTimer::TimeAccumulator::reset() {
+  elapsed_s = 0.0;
+  number_of_entrances = 0;
+}
+
+/** Increment the elapsed wall-clock time and number of times the code block was entered
+ * @param time_s :: elapsed time (seconds) to add
+ */
+void CodeBlockMultipleTimer::TimeAccumulator::increment(const double time_s) {
```

Comment:
I haven't seen usage, but would it simplify clients by adding an overload that takes two timestamps (i.e. `std::chrono::system_clock::now()`)?

Reply:
It is true that, in principle, time calculation may be implemented in either `CodeBlockMultipleTimer` or `CodeBlockMultipleTimer::TimeAccumulator`. Still, I think it's a cleaner design when the timer calculates the elapsed time after each code block execution, and the time accumulator only keeps track of the total time spent in all calls.